Adds ability to set custom html attributes in Panels

pySilver opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

This is the take-over original #10323 PR.

Fixes https://github.com/wagtail/wagtail/issues/10133

-   [x] Synced with the current main branch state
-   [x] Fixes test
-   [x] Battle-tested and reworked logic of attributes placement, a few details below.

## Differences in this PR
- It is not always possible to render html attributes directly to the Panel's template. Some panels simply do not have any wrapping HTML node, such as `ObjectList`, `FieldRowPanel`, or `MultiFieldPanel` for this use case we render attributes in the closest common parent. Right, where the developer would expect it to be.
- `Panel.BASE_ATTRS` is not used in the child panels at the moment since their data attributes are expected to be rendered somewhere else and it is not possible to cover both cases. However, this attribute can be used in future panel refactoring or currently when the panel gets extended by the custom code.
- Wagtail's data attributes stayed untouched in the original templates so UI will stay functional after the merge.
- I've tested the presented approach while integrating my own Stimulus controllers and found no issues. I am able to set specific attributes on any required level. (I wasn't able to achieve that in original PR)






_Please check the following:_

-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] Do the tests still pass?[^1]
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [-] For new features: Has the documentation been updated accordingly?

pySilver commented:
I'm not sure what this failing template check means and how to mitigate that. I did run `make lint` before publishing PR.

https://app.circleci.com/pipelines/github/wagtail/wagtail/14149/workflows/d12a95c9-0b00-43d9-8b02-6990f6785932/jobs/40247/parallel-runs/0/steps/0-108

lb- reviewed (changes requested):
@pySilver  thank you for getting to this, I think the approach of not using `BASE_ATTRS` for any of the current data attributes (tabs, data-panel etc) makes sense.

When I use this branch though I am seeing the attrs being duplicated for any nested panels.

e.g. a `HelpPanel` or `FieldPanel` within `MultiFieldPanel` or `FieldRowPanel` gets repeated on the outer wrapper and the field itself.

I think this use case will need tests, so that we can confirm standalone panels within nested panels have the right attributes and visa-versa.

Additionally, are you able to review the original PR for feedback about documentation, it would be good to have this added also.

This is close but still needs some refinements.

## Example

```py
        MultiFieldPanel(
            [
                # Example use case for HelpPanel.
                HelpPanel(
                    "Refer to keywords analysis and correct international ingredients names to craft the best introduction backstory, and headline.",
                    attrs={"data-field-test": "help-panel"},
                ),
```

See below the `data-field-test` is on two elements.
![Screenshot 2023-07-06 at 3 15 47 pm](https://github.com/wagtail/wagtail/assets/1396140/4be26a75-83e5-463b-95b8-e10b8f9077bb)


Thank you for pursuing this.

pySilver commented on wagtail/admin/tests/test_edit_handlers.py:
```diff
@@ -1622,7 +1640,6 @@ def test_form(self):
         self.assertEqual(replies_formset.forms[0].for_user, self.commenting_user)
 
     def test_comment_form_validation(self):
-
```
Oh. I'm not sure where it comes from. I've applied `make format-server` before comitting.

lb- replied:
Hmm ok. Maybe it's needed. Thanks for checking

pySilver commented:
@lb- I've made a few changes:

1. Added a separate test case for a fairly complex panel layout. It also checks that a particular attribute is only applied once.
2. I've had to remove direct attributes support at `FieldRowPanel` as it is not needed as long as it's a part of some other panel (usually `ObjectList` or `MultiFieldPanel`)
3. I've added documentation for the Panel with an example of Stimulus controller integration.
4. Linted & Formatted.

pySilver commented:
@lb- I'm not sure what's wrong with this failing test. I believe it has nothing to do with my code. I'm running a test suite locally and it works just fine.

lb- commented:
@pySilver don't worry about that single failing test, it's running against Django main to flag any future compatibility issues. It's not a problem for this PR.

lb- commented:
I'll label as needs review, I'm away for a bit so I won't get to it until late next week. Someone else on the core team may review or add comments in the meantime. 

Thanks for working through the issues @pySilver

lb- reviewed (approved):
Tested this on each kind of Panel combination I could think of and it's looking solid! thank you so much @pySilver 
Also thanks @salty-ivy  for the original work on this.

I have updated some documentation on my local version and will rebase/squash etc shortly.

Approved.

lb- commented:
Note: I pulled in some of the test cases from the original PR also, it had some extra bolstering of this feature plus checks that the admin required data attributes are included.
